src/transformStateWithClones.js

```diff
@@ -6,8 +6,42 @@
  *
  * @return {Object[]}
  */
+const addProperties = 'addProperties';
+const removeProperties = 'removeProperties';
+const clear = 'clear';
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const states = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    let nextState;
+
+    switch (action.type) {
+      case clear:
+        nextState = {};
+        break;
+
+      case addProperties:
+        nextState = { ...currentState, ...action.extraData };
+        break;
+      case removeProperties:
+        nextState = { ...currentState };
+
+        for (const key of action.keysToRemove) {
+          delete nextState[key];
+        }
+        break;
+
+      default:
+        continue;
```

Comment:
```suggestion
        throw new Error(`Invalid action type: ${action.type}`);
```
It's better to throw an error in case someone passes invalid action